client-js/app/scripts/app.ts

```diff
@@ -141,54 +141,55 @@ window.addEventListener('WebComponentsReady', () => {
 
   class RegistryActivity implements Activity {
     path: Array<string>;
+    private elem: LabradRegistry;
 
-    constructor(path: Array<string>) {
+    private api: registry.RegistryApi;
+    private lifetime = new Lifetime();
+
+    constructor(api: registry.RegistryApi, path: Array<string>) {
+      this.api = api;
       this.path = path;
     }
 
     start(): Promise<ActivityState> {
+      this.api.newItem.add(() => this.onNewItem(), this.lifetime);
       console.log('loading registry:', this.path);
-      return reg.dir({path: this.path}).then((listing) => {
-        var breadcrumbs = [];
-        for (var i = 0; i <= this.path.length; i++) {
-          breadcrumbs.push({
-            name: (i == 0) ? 'registry' : this.path[i-1],
-            isLink: i < this.path.length,
-            url: registryUrl(this.path.slice(0, i))
-          });
-        }
+      return this.api.watch({path: this.path}).then(() => {
+        return reg.dir({path: this.path}).then((listing) => {
```

Comment:
One of the nice things about promises is that you can avoid levels of nesting and instead string `.then` callbacks together sequentially. Here, for example, you can do:
```typescript
return this.api.watch({path: this.path})
  .then(() => this.api.dir({path: this.path}))
  .then((listing) => {
    var breadcrumbs = [];
    ...
  });
```
I think this will be even nicer when we can use async functions, as I mentioned previously, but as long as we're using explicit promises I think the sequential composition is clearer than nesting.